src/gt4py/next/program_processors/runners/dace_fieldview/transformations/util.py

```diff
@@ -0,0 +1,168 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""Common functionality for the transformations/optimization pipeline."""
+
+from typing import Iterable, Union
+
+import dace
+from dace.sdfg import graph as dace_graph, nodes as dace_nodes
+
+
+def is_nested_sdfg(
+    sdfg: Union[dace.SDFG, dace.SDFGState, dace_nodes.NestedSDFG],
+) -> bool:
+    """Tests if `sdfg` is a NestedSDFG."""
+    if isinstance(sdfg, dace.SDFGState):
+        sdfg = sdfg.parent
+    if isinstance(sdfg, dace_nodes.NestedSDFG):
+        return True
+    elif isinstance(sdfg, dace.SDFG):
+        if sdfg.parent_nsdfg_node is not None:
+            return True
+        return False
+    else:
+        raise TypeError(f"Does not know how to handle '{type(sdfg).__name__}'.")
+
+
+def all_nodes_between(
+    graph: dace.SDFG | dace.SDFGState,
+    begin: dace_nodes.Node,
+    end: dace_nodes.Node,
+    reverse: bool = False,
+) -> set[dace_nodes.Node] | None:
+    """Find all nodes that are reachable from `begin` but bound by `end`.
+
+    Essentially the function starts a DFS at `begin`. If an edge is found that lead
+    to `end`, this edge is ignored. It will thus found any node that is reachable
+    from `begin` by a path that does not involve `end`. The returned set will
+    never contain `end` nor `begin`. In case `end` is never found the function
+    will return `None`.
+
+    If `reverse` is set to `True` the function will start exploring at `end` and
+    follows the outgoing edges, i.e. the meaning of `end` and `begin` are swapped.
+
+    Args:
+        graph: The graph to operate on.
+        begin: The start of the DFS.
+        end: The terminator node of the DFS.
+        reverse: Perform a backward DFS.
+
+    Notes:
+        - The returned set will also contain the nodes of path that starts at
+            `begin` and ends at a node that is not `end`.
+    """
+
+    def next_nodes(node: dace_nodes.Node) -> Iterable[dace_nodes.Node]:
+        if reverse:
+            return (edge.src for edge in graph.in_edges(node))
+        return (edge.dst for edge in graph.out_edges(node))
+
+    if reverse:
+        begin, end = end, begin
+
+    to_visit: list[dace_nodes.Node] = [begin]
+    seen: set[dace_nodes.Node] = set()
+    found_end: bool = False
+
+    while len(to_visit) > 0:
+        n: dace_nodes.Node = to_visit.pop()
+        if n == end:
+            found_end = True
+            continue
+        elif n in seen:
+            continue
+        seen.add(n)
+        to_visit.extend(next_nodes(n))
+
+    if not found_end:
+        return None
+
+    seen.discard(begin)
+    return seen
```

Comment:
Just a suggestion to make the code simpler:
```suggestion
    while len(to_visit) > 0:
        n: dace_nodes.Node = to_visit.pop()
        if n not in seen:
            seen.add(n)
            if n != end:
                to_visit.extend(next_nodes(n))

    return seen - {begin, end} if end in seen else None
```

Reply:
I liked your idea and applied it.